math/Fields.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include <cstddef>
+#include <type_traits>
+
+namespace math
+{
+
+// difference (b - a) mod (1<<BITS)
+template <typename T, size_t BITS>
+constexpr typename std::make_signed<T>::type ringDifference(T a, const T b)
+{
+    static_assert(std::is_unsigned<T>::value, "Difference only works on unsigned integers");
+    using SignedT = typename std::make_signed<T>::type;
+    if (a & (1 << (BITS - 1)))
```

Comment:
@RicardoMDomingues , as far as I understood this should work _only_ with smaller "rings". E.g. 63 bits and smalller. So in your example 'a' could be uint32_t.

Reply:
Not sure if understood. `a` can be a `uint64_t` and BITS to be 63? That will also cause issues. The expression `(1 << (BITS - 1))` will cause a compiler warning each time `BITS >= 32` as shift count is bigger than number of int bits. So in the end it will be truncated to int before be converted to uint64_t

Reply:
don't worry I am working on it. I also see cases were it is not producing what I expect.

Reply:
What I meant is that value of `a` should be expressed by BITS number of bits. So in your example value of `a` should be 0x000000007FFFFFFF.

Reply:
Even if  this work for our current cases this accept a generic types. One day someone could find this function useful and use with a combination hat does not work correctly and it will create problem that are very hard to find